storage: optimistically resolve pending prereqs in cmdQ before canceling

nvb opened the pull request:
In a privately reported issue we have seen cases where cascading
cancellations of commands in the CommandQueue can cause an OOM
in `storage.(*cmd).ResolvePendingPrereq`. This is caused by a quadratic
blowup of prerequisites as canceled commands transfer their prerequisites
to their dependents.

This change introduces a measure to bound this quadratic blowup by
adding in a new optimistic resolution step to command cancellation,
during which time a command will optimistically resolve all prerequisites
that have already finished without blocking on any pending commands. This
should improve the cascading cancellation scenario because it will
allow a command to trim its prerequisite list before passing it on
to its dependents. Notably, since we consider all commands in the
prereq list instead of just the first, like we usually do, a single
pending command at the front of the list can't allow the rest of the
prereq list to grow full of finished/canceled commands.

Release note: None

nvb committed:
storage: optimistically resolve pending prereqs in cmdQ before canceling

In a privately reported issue we have seen cases where cascading
cancellations of commands in the CommandQueue can cause an OOM
in `storage.(*cmd).ResolvePendingPrereq`. This is caused by a quadratic
blowup of prerequisites as canceled commands transfer their prerequisites
to their dependents.

This change introduces a measure to bound this quadratic blowup by
adding in a new optimistic resolution step to command cancellation,
during which time a command will optimistically resolve all prerequisites
that have already finished without blocking on any pending commands. This
should improve the cascading cancellation scenario because it will
allow a command to trim its prerequisite list before passing it on
to its dependents. Notably, since we consider all commands in the
prereq list instead of just the first, like we usually do, a single
pending command at the front of the list can't allow the rest of the
prereq list to grow full of finished/canceled commands.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/27117)
<!-- Reviewable:end -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
Is there a way to set up a unit test that would previously exhibit the quadratic growth? This would be useful mostly to get an idea of how deep a dependency tree we were looking at in this case.

---

Reviewed 2 of 2 files at r1.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/storage/replica.go, line 2324 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-LGTT47c1A2-QwClqHs0:-LGTT47dCAZJ5G9XTqFb:b-qhz2yl) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3188c51d4bdaf3d4fafaab3a6f00c29731b2c8/pkg/storage/replica.go#L2324)):*
> ```Go
> 						// Remove the command from the command queue immediately. New commands that
> 						// would have established a dependency on this command will never see it,
> 						// which is fine. Current dependents  that already have a dependency on this
> ```

monster nit: double space.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-:-LGTTF-r3U6VqoWMqwX8:b-yh5mxc)*
<!-- Sent from Reviewable.io -->

nvb committed:
storage: avoid migrating duplicate prereq commands in cmdQ

We now avoid transferring duplicate prereq commands from
a canceled prereq to a dependent command in the CommandQueue.
This avoids quadratic growth of memory when a large number
of commands are canceled which each share a command
prerequisite. Previously, this could result in a memory
explosion if it happened repeatedly.

The commit also adds a test that demonstrates the
effect of this change and the previous to optimistically
resolve pending prereqs in the cmdQ before canceling.

Release note: None

nvb reviewed (commented):
I added a second commit that addresses an even more serious source of quadratic growth and introduces a test which stresses cascading command cancellation. The test times out without these changes because it gets so slow. I also added in instrumentation when building it which confirmed that the largest prereq slices were growing huge without these fixes. With them, the prereq slice never grows larger than `O(# of pending cmds)`. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-:-LGTiKoaDxla_BbMps3z:b-myrtj2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/replica.go, line 2324 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-LGTT47c1A2-QwClqHs0:-LGTkCL5BWZAze_9BJQA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3188c51d4bdaf3d4fafaab3a6f00c29731b2c8/pkg/storage/replica.go#L2324)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

monster nit: double space.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg reviewed (changes requested):
Really nice, in particular the test! <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-:-LIKabXv1Vo6RPlPeJ24:b-a9ta2p)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/storage/command_queue.go, line 243 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-LIKWKKG7l1NXEA2X1-B:-LIKWKKG7l1NXEA2X1-C:be1o39a) ([raw file](https://github.com/cockroachdb/cockroach/blob/728ca7324c5205cfad8eebca90c14f1c04ec062b/pkg/storage/command_queue.go#L243)):*
> ```Go
> 		if c.prereqIDs == nil {
> 			// Lazily compute prereq ID set. This is only necessary during
> 			// command cancellation scenerios.
> ```

scenarios
___
*[pkg/storage/command_queue.go, line 265 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-LIKWxwr0VSSya9NPq8H:-LIKWxwr0VSSya9NPq8I:bwz4oyp) ([raw file](https://github.com/cockroachdb/cockroach/blob/728ca7324c5205cfad8eebca90c14f1c04ec062b/pkg/storage/command_queue.go#L265)):*
> ```Go
> 	(*c.prereqs) = (*c.prereqs)[1:]
> 
> 	// Delete from the prereq ID set, if initialized.
> ```

I stumbled over this, perhaps change to

// Delete from the prereq ID set (if c.prereqIDs is nil, this is a no-op).
___
*[pkg/storage/replica_test.go, line 2525 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-LIKWdKT5kV2pENl1z7R:-LIKWdKT5kV2pENl1z7S:bnd6qad) ([raw file](https://github.com/cockroachdb/cockroach/blob/728ca7324c5205cfad8eebca90c14f1c04ec062b/pkg/storage/replica_test.go#L2525)):*
> ```Go
> 	//         ...
> 	//
> 	// The test then creates a cascade cancellation scenerio
> ```

scenario
___
*[pkg/storage/replica_test.go, line 2547 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-LIKYEKnCEqIyO1dY3F3:-LIKYEKnCEqIyO1dY3F4:bnd6qad) ([raw file](https://github.com/cockroachdb/cockroach/blob/728ca7324c5205cfad8eebca90c14f1c04ec062b/pkg/storage/replica_test.go#L2547)):*
> ```Go
> 	//    dependent commands.
> 	// 2. Alternating multi-command layers and single command layers creates
> 	//    a scenerio where dependencies would add duplicate prereqs to their
> ```

scenario


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-:-LILZP5YAhzK9DpAj9mC:b-814g8k)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/storage/command_queue.go, line 243 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-LIKWKKG7l1NXEA2X1-B:-LILZ8Y_AQkGqsXByPQe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/728ca7324c5205cfad8eebca90c14f1c04ec062b/pkg/storage/command_queue.go#L243)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

scenarios
</blockquote></details>

Done.
___
*[pkg/storage/command_queue.go, line 265 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-LIKWxwr0VSSya9NPq8H:-LILZODb8PWERUMjh0EB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/728ca7324c5205cfad8eebca90c14f1c04ec062b/pkg/storage/command_queue.go#L265)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I stumbled over this, perhaps change to

// Delete from the prereq ID set (if c.prereqIDs is nil, this is a no-op).
</blockquote></details>

Done.
___
*[pkg/storage/replica_test.go, line 2525 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-LIKWdKT5kV2pENl1z7R:-LILZ96X6QtlfjNMHsui:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/728ca7324c5205cfad8eebca90c14f1c04ec062b/pkg/storage/replica_test.go#L2525)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

scenario
</blockquote></details>

Done.
___
*[pkg/storage/replica_test.go, line 2547 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-LIKYEKnCEqIyO1dY3F3:-LILZAUI1ak0E8G4GmjC:b-jn2qhb) ([raw file](https://github.com/cockroachdb/cockroach/blob/728ca7324c5205cfad8eebca90c14f1c04ec062b/pkg/storage/replica_test.go#L2547)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

scenario
</blockquote></details>

Done. You sure there aren't any scenerios where this is right? :trollface: 


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

btw, is there a magical trick to approve a PR even if you have comments open? I guess I could dismiss myself, but that seems awkward.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27117#-:-LILaLf5-6JyESqrGZsI:b-s4lgos)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
TFTR!

bors r=tschottdorf
